### PR TITLE
SSH verification

### PR DESCRIPTION
**Pull request by ameade**
Added ssh verification to the test suite and implemented verification in reboot tests

### CODE REVIEW

File: stacktester/common/http.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ class Client(object):
 
     def poll_request(self, method, url, check_response, **kwargs):
 
-        timeout = kwargs.pop('timeout', 180)
+        timeout = kwargs.pop('timeout', 300)
```
Comment (bcwaldon) [inline stacktester/common/http.py]: What is the reasoning for this?

File: stacktester/common/ssh.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                time.sleep(5)
+            ssh.close()
+        except EOFError:
+            return
+        except paramiko.AuthenticationException:
+            return
+        except socket.error:
+            return
+
+    def get_time_started(self, host, username, password):
```
Comment (bcwaldon) [inline stacktester/common/ssh.py]: I think this function should live somewhere Nova-specific. Let's keep the common.ssh.Client class for common functionality.

File: stacktester/common/ssh.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class Client(object):
+
+
+    def __init__(self, host='localhost', port=80, base_url=''):
+        self.config = stacktester.config.StackConfig()
+        self.ssh_timeout = self.config.nova.ssh_timeout
+
+    def get_ssh_connection(self, host, user, pwd):
```
Comment (bcwaldon) [inline stacktester/common/ssh.py]: It is fine to use this kind of method internally, but we shouldn't expose it as a part of our interface. You typically want your Client implementation to handle all of the connection-level stuff, and not return the actual connection object to the user. How about you go for a connect/disconnect interface where the ssh connection object is used only within the Client class. Then you can add an exec_command method to this class that just calls the local ssh connection method.

File: stacktester/common/ssh.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
+    import paramiko
+
+
+class Client(object):
+
+
+    def __init__(self, host='localhost', port=80, base_url=''):
```
Comment (bcwaldon) [inline stacktester/common/ssh.py]: Looks like somebody copied this code from somewhere else :) Let's remove base_url.

File: stacktester/common/ssh.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        self.config = stacktester.config.StackConfig()
+        self.ssh_timeout = self.config.nova.ssh_timeout
```
Comment (bcwaldon) [inline stacktester/common/ssh.py]: You shouldn't read from the config inside this object. Pass in the timeout as a kwarg.

File: stacktester/tests/test_server_actions.py [hunk 6]
```diff
@@ -44,10 +46,20 @@ class ServerActionsTest(unittest.TestCase):
 
         self.server_id = data['server']['id']
         self.os.nova.wait_for_server_status(self.server_id, 'ACTIVE')
+        
+        response, body = self.os.nova.request(
+            'GET', '/servers/%s' % self.server_id, body=post_body)
+        #self.assertEqual('200', response['status'])
+
+        data = json.loads(body)
+        #current impl
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py:55]: As a convention, we have been using KNOWN-ISSUE to mark these situations

File: stacktester/tests/test_server_actions.py [hunk 7]
```diff
@@ -56,7 +68,9 @@ class ServerActionsTest(unittest.TestCase):
         })
         response, body = self.os.nova.request(
             'DELETE',
-            '/servers/%s' % self.server_id)
+            '/servers/%s' % self.server_id,
+            body=delete_body)
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: DELETE actions don't need a body.

File: stacktester/tests/test_server_actions.py [hunk 8]
```diff
@@ -69,40 +83,108 @@ class ServerActionsTest(unittest.TestCase):
...
+        self.ssh.connect_until_closed(self.access_ip, 
+                                      'root',
+                                      'testpwd')
+        #ssh and verify uptime is less than before
+        post_reboot_time_started = self.ssh.get_time_started(self.access_ip, 
+                                                        'root',
+                                                        'testpwd')
+        self.assertTrue(initial_time_started < post_reboot_time_started)
 
-        #verify state change
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py:79]: What is the reason for removing this?

File: stacktester/tests/test_server_actions.py [hunk 9]
```diff
@@ -69,40 +83,108 @@ class ServerActionsTest(unittest.TestCase):
...
+            'server' : {
+                'name' : 'testserver',
+                'imageRef' : self.image_ref,
+                'flavorRef' : self.flavor_ref,
+            }
+        })
+        response, body = self.os.nova.request(
+            'DELETE',
+            '/servers/%s' % self.server_id,
+            body=delete_body)
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: No body required for DELETE.

File: stacktester/tests/test_server_actions.py [hunk 10]
```diff
@@ -69,40 +83,108 @@ class ServerActionsTest(unittest.TestCase):
...
+        self.assertTrue(initial_time_started < post_reboot_time_started)
 
-        #verify state change
-        # KNOWN-ISSUE lp?         
-        #self._wait_for_status(self.server_id, 'HARD_REBOOT')
-        self._wait_for_status(self.server_id, 'ACTIVE')
+
+class ServerActionsTest(unittest.TestCase):
+
+    multi_node = openstack.Manager().config.env.multi_node
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: Doing it this way creates an openstack.Manager() object and throws it away immediately. You should set this in setUpClass.

Comment (dwalleck) [inline stacktester/tests/test_server_actions.py]: The issue here is that this config has to be made outside of the setup as the skip decorator doesn't have a reference to self. That is unless there's a python-esque way to do this that I haven't seen.

Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: Okay, I didn't notice that. Could we just use a config object, rather than a whole manager?

File: stacktester/common/ssh.py (hunk 3 above)
Comment (throughnothing) [inline stacktester/common/ssh.py]: +1

File: stacktester/tests/test_server_actions.py (hunk 9 above)
Comment (throughnothing) [inline stacktester/tests/test_server_actions.py]: Still think we need to get rid of this body here, then it looks good.

File: stacktester/tests/test_server_actions.py [hunk 11]
```diff
@@ -98,11 +119,57 @@ class ServerActionsTest(unittest.TestCase):
...
         self.assertEqual(response['status'], '202')
+        self.os.nova.wait_for_server_status(self.server_id, 'ACTIVE')
+        self.ssh.connect_until_closed()
+        #ssh and verify uptime is less than before
+        post_reboot_time_started = self._get_time_started()
 
-        #verify state change
-        # KNOWN-ISSUE lp?         
-        #self._wait_for_status(self.server_id, 'HARD_REBOOT')
-        self._wait_for_status(self.server_id, 'ACTIVE')
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: Why was this removed?

File: stacktester/tests/test_server_actions.py [hunk 12]
```diff
@@ -64,11 +72,20 @@ class ServerActionsTest(unittest.TestCase):
         except exceptions.TimeoutException:
             self.fail("Server failed to change status to %s" % status)
 
+    def _get_time_started(self):
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: This name is a little confusing to me. Without looking at the implementation, I first thought it was the created_at timestamp in the API. Could we change this to get_boot_time?

File: stacktester/tests/test_server_actions.py [hunk 13]
```diff
@@ -78,17 +95,21 @@ class ServerActionsTest(unittest.TestCase):
         response, body = self.os.nova.request(
             'POST', "/servers/%s/action" % self.server_id, body=post_body)
         self.assertEqual(response['status'], '202')
+        self.os.nova.wait_for_server_status(self.server_id, 'ACTIVE')
+        self.ssh.connect_until_closed()
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py:92]: I think I would like to depend on api status changes instead of ssh connections. Anybody have any thoughts? If we want to keep the current behavior, I think changing the name to something like block_while_available or something like that might help communicate what is happening.

File: stacktester/tests/test_server_actions.py [hunk 14]
```diff
@@ -98,11 +119,57 @@ class ServerActionsTest(unittest.TestCase):
...
+        post_body = json.dumps({
+            'server' : {
+                'name' : 'testserver',
+                'imageRef' : self.image_ref,
+                'flavorRef' : self.flavor_ref,
+                'adminPass' : "testpwd",
+            }
+        })
+
+        response, body = self.os.nova.request(
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: You can use the create_server helper on the self.os.nova object. Same goes for delete_server.

File: stacktester/tests/test_server_actions.py [hunk 15]
```diff
@@ -13,28 +13,30 @@
...
+from stacktester.common import ssh
 
 import unittest2 as unittest
 
 
-class ServerActionsTest(unittest.TestCase):
-    
-    multi_node = openstack.Manager().config.env.multi_node
- 
+class ServerRebootActionTest(unittest.TestCase):
```
Comment (bcwaldon) [inline stacktester/tests/test_server_actions.py]: I would like to either do all of the actions in individual test cases, or do them all in one.

File: stacktester/common/ssh.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+class Client(object):
+
+
+    def __init__(self, host, username, password, timeout=300):
+        self.host = host
+        self.username = username
+        self.password = password
+        self.timeout = timeout
+
+    def _ssh_connection(self):
```
Comment (bcwaldon) [inline stacktester/common/ssh.py]: Can you call this _get_ssh_connection? I like seeing method names written as actions. This makes it seem like an attribute.

File: stacktester/tests/test_server_actions.py (hunk 13 above)
Comment (ameade) [inline stacktester/tests/test_server_actions.py:92]: We can't depend on the api status here because it will get to that next line before the instance status has been set to REBOOT,
